src/hotspot/share/gc/g1/g1ConcurrentMark.cpp

```diff
@@ -695,10 +695,16 @@ void G1ConcurrentMark::pre_concurrent_start() {
   _g1h->heap_region_iterate(&startcl);
 
   _root_regions.reset();
+
+  _gc_tracer_cm->set_gc_cause(cause);
 }
 
 
-void G1ConcurrentMark::post_concurrent_start() {
+void G1ConcurrentMark::post_concurrent_start(bool concurrent_operation_is_full_mark) {
+  if (!concurrent_operation_is_full_mark) {
+    root_regions()->cancel_scan();
+    return;
+  }
```

Comment:
I still think this should be broken out from this function. Either doing it in `G1ConcurrentMarkThread::concurrent_undo_cycle_do()` or having a separate function that is called right before `_cm_thread->start_undo_mark()` in `CollectedHeap::start_concurrent_cycle()`.

Reply:
Done in the latest change.